### PR TITLE
[SPARK-38270][SQL] Spark SQL CLI's AM should keep same exit code with client side

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
Current Spark SQL CLI alway use  shutdown hook to stop SparkSQLEnv
```
 // Clean up after we exit
    ShutdownHookManager.addShutdownHook { () => SparkSQLEnv.stop() }
```

but use process ret to close client side jvm
```
while (line != null) {
  if (!line.startsWith("--")) {
    if (prefix.nonEmpty) {
      prefix += '\n'
     } 
    if (line.trim().endsWith(";") && !line.trim().endsWith("\\;")) {
      line = prefix + line
      ret = cli.processLine(line, true)
      prefix = ""
      currentPrompt = promptWithCurrentDB
    } else {
      prefix = prefix + line
      currentPrompt = continuedPromptWithDBSpaces
    }
  }
  line = reader.readLine(currentPrompt + "> ")
}

sessionState.close()

System.exit(ret)
}
```

```
    if (sessionState.execString != null) {
      exitCode = cli.processLine(sessionState.execString)
      System.exit(exitCode)
    }

    try {
      if (sessionState.fileName != null) {
        exitCode = cli.processFile(sessionState.fileName)
        System.exit(exitCode)
      }
    } catch {
      case e: FileNotFoundException =>
        logError(s"Could not open input file for reading. (${e.getMessage})")
        exitCode = 3
        System.exit(exitCode)
    }

```

This always cause client side exit code not consistent with AM.

IN this pr I prefer to pass the exit code to `SparkContext.stop()` method to pass a clear client side status to AM side in client mode. 

In this pr, I add a new `stop` method in `SchedulerBackend`
```
def stop(exitCode: Int): Unit = stop()
```
So we don't need to implement it for all kinds of scheduler backend.
In this pr, we only handle the case me meet for `YarnClientSchedulerBackend`, then we can only implement `stop(exitCode: Int)` for this class, then for yarn client mode, it can work now.

I think this can benefit many similar case in future.

### Why are the changes needed?
Keep client side status consistent  with AM side

### Does this PR introduce _any_ user-facing change?
With this pr, client side status will be same with am side

### How was this patch tested?
MT

### CODE REVIEW

**Comment by LuciferYang** [comment]
> IN this pr I prefer to pass the exit code to SparkContext.stop() method to pass a clear cliet side status to AM side in client mode. And I think this can benefit many similar case in future.

cliet -> client

**Comment by AngersZhuuuu** [comment]
> > IN this pr I prefer to pass the exit code to SparkContext.stop() method to pass a clear cliet side status to AM side in client mode. And I think this can benefit many similar case in future.
> 
> cliet -> client

updated

**Comment by AngersZhuuuu** [comment]
gentle ping @cloud-fan @tgravescs @mridulm

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 1]
```diff
@@ -2059,7 +2059,9 @@ class SparkContext(config: SparkConf) extends Logging {
   /**
    * Shut down the SparkContext.
    */
-  def stop(): Unit = {
+  def stop(): Unit = stop(0)
+
+  def stop(exitCode: Int): Unit = {
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/SparkContext.scala:2064]: This is a public API and we need to document it properly.

**Comment by cloud-fan** [comment]
cc @bogdanghit

File: core/src/main/scala/org/apache/spark/SparkContext.scala (hunk 1 above)
Comment (AngersZhuuuu) [reply core/src/main/scala/org/apache/spark/SparkContext.scala:2075]: > This is a public API and we need to document it properly.

Updated

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala [hunk 2]
```diff
@@ -846,8 +850,13 @@ private[spark] class ApplicationMaster(
       // This avoids potentially reporting incorrect exit codes if the driver fails
       if (!(isClusterMode || sparkConf.get(YARN_UNMANAGED_AM))) {
         if (shutdown || !clientModeTreatDisconnectAsFailed) {
-          logInfo(s"Driver terminated or disconnected! Shutting down. $remoteAddress")
-          finish(FinalApplicationStatus.SUCCEEDED, ApplicationMaster.EXIT_SUCCESS)
+          if (exitCode == 0) {
+            logInfo(s"Driver terminated or disconnected! Shutting down. $remoteAddress")
+            finish(FinalApplicationStatus.SUCCEEDED, ApplicationMaster.EXIT_SUCCESS)
+          } else {
+            logInfo(s"Driver terminated with exit code ${exitCode}! Shutting down. $remoteAddress")
```
Comment (yaooqinn) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala:857]: logError

Comment (AngersZhuuuu) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: > logError

DOne

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 3]
```diff
@@ -2059,7 +2059,20 @@ class SparkContext(config: SparkConf) extends Logging {
...
+  def stop(): Unit = stop(0)
+
+  /**
+   * Shut down the SparkContext with exit code that will passed to scheduler backend.
+   * In client mode, client side may call `SparkContext.stop()` to clean up but exit with
+   * code not equal to 0. This behavior cause resource scheduler such as `ApplicationMaster`
+   * exit with success status but client side exited with failed status. Spark can call
+   * this method to stop SparkContext and pass client side correct exit code to scheduler backend.
+   * Then scheduler backend should send the exit code to corresponding resource scheduler
+   * to keep consistens.
```
Comment (yaooqinn) [inline core/src/main/scala/org/apache/spark/SparkContext.scala:2071]: consistent

Comment (AngersZhuuuu) [reply core/src/main/scala/org/apache/spark/SparkContext.scala:2071]: Done

**Review by yaooqinn — APPROVED** [review]

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIDriver.scala [hunk 4]
```diff
@@ -139,7 +145,7 @@ private[hive] object SparkSQLCLIDriver extends Logging {
     SessionState.setCurrentSessionState(sessionState)
 
     // Clean up after we exit
-    ShutdownHookManager.addShutdownHook { () => SparkSQLEnv.stop() }
+    ShutdownHookManager.addShutdownHook { () => SparkSQLEnv.stop(exitCode) }
```
Comment (LuciferYang) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIDriver.scala:148]: ShutdownHook is executed on another thread,  `exitCode` should be `volatile`

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIDriver.scala [hunk 5]
```diff
@@ -476,7 +482,7 @@ private[hive] class SparkSQLCLIDriver extends CliDriver with Logging {
           // Kill the VM on second ctrl+c
           if (!initialRequest) {
             console.printInfo("Exiting the JVM")
-            System.exit(127)
+            SparkSQLCLIDriver.exit(127)
```
Comment (LuciferYang) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIDriver.scala:485]: Although not introduced by this pr, but I think we should give `127` a definite name.

Comment (AngersZhuuuu) [reply sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIDriver.scala:485]: > Although not introduced by this pr, but I think we should give `127` a definite name.

I will raise a new pr for all CLI's exit code.

**Review by LuciferYang — APPROVED** [review]
LGTM +1

**Comment by AngersZhuuuu** [comment]
gentle ping @cloud-fan @bogdanghit

**Comment by bogdanghit** [comment]
@AngersZhuuuu does this address the CliSuite flakiness I reported a while back?

**Comment by AngersZhuuuu** [comment]
> @AngersZhuuuu does this address the CliSuite flakiness I reported a while back?

Can you provide the link？  Literally, it should be irrelevant

**Comment by AngersZhuuuu** [comment]
gentle ping @bogdanghit

**Comment by cloud-fan** [comment]
@AngersZhuuuu can you rebase? We should merge this PR.

**Comment by yaooqinn** [comment]
+1, and sorry for not merging it after my approval

**Comment by AngersZhuuuu** [comment]
> @AngersZhuuuu can you rebase? We should merge this PR.

Done

**Review by yaooqinn — APPROVED** [review]

**Comment by AngersZhuuuu** [comment]
ping @cloud-fan @yaooqinn @LuciferYang

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**